File: web/src/lib/aos-fact-extract.ts
/**
 * AOS discretionary brief — extract structured drafting facts from attorney paste summary.
 * Heuristic path runs client-side; LLM path via POST /api/aos/extract-facts → Fly API.
 */

import {
  emptyDraftingFacts,
  type DraftingFactsPayload,
} from "./practice-area-facts";
import { mergeOcrIntoDraftingFacts, type OcrFact } from "./intake-prefill";

export type AosExtractResult = {
  fields: Record<string, string>;
  paragraphSelections: Record<string, string>;
  additionalNotes: string;
  confidence: "high" | "medium" | "low";
  extractionMode: "llm" | "heuristic";
  llmAvailable: boolean;
  llmFallback?: boolean;
};

const AOS_FIELD_IDS = [
  "applicantName",
  "aNumber",
  "clientStatus",
  "entryDate",
  "portOfEntry",
  "entryVisaType",
  "petitionerName",
  "petitionerRelationship",
  "qualifyingRelative",
  "i130ApprovedDate",
  "i485FiledDate",
  "reliefSought",
  "caseTheme",
  "caseThemeBrief",
  "sectionAHeading",
  "sectionAFacts",
  "sectionBHeading",
  "sectionBFacts",
  "adverseHeading",
  "adverseFactorBrief",
  "positiveEquities",
  "balancingInventory",
  "adverseFacts",
  "adverseFactors",
  "departureHarm",
  "extremeHardshipFactors",
  "inadmissibilityGrounds",
  "priorFilings",
] as const;

const DATE_PATTERNS = [
  /\b(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\.?\s+\d{1,2},?\s+\d{4}\b/gi,
  /\b\d{1,2}[/-]\d{1,2}[/-]\d{2,4}\b/g,
  /\b\d{4}-\d{2}-\d{2}\b/g,
];

function firstDateNear(text: string, keywords: string[]): string {
  const lower = text.toLowerCase();
  for (const kw of keywords) {
    const idx = lower.indexOf(kw);
    if (idx < 0) continue;
    const window = text.slice(Math.max(0, idx - 80), idx + kw.length + 120);
    for (const pattern of DATE_PATTERNS) {
      pattern.lastIndex = 0;
      const match = pattern.exec(window);
      if (match) return match[0].trim();
    }
  }
  return "";
}

function matchGroup(pattern: RegExp, text: string): string {
  const match = text.match(pattern);
  return match?.[1]?.trim() ?? "";
}

/** Deterministic extract — no API key required. */
export function heuristicExtractAosFacts(summary: string): AosExtractResult {
  const text = summary.trim();
  const fields: Record<string, string> = Object.fromEntries(AOS_FIELD_IDS.map((id) => [id, ""]));

  if (!text) {
    return {
      fields,
      paragraphSelections: {},
      additionalNotes: "",
      confidence: "low",
      extractionMode: "heuristic",
      llmAvailable: false,
    };
  }

  const aNum = text.match(/\bA-?\d{8,9}\b/i);
  if (aNum) fields.aNumber = aNum[0].toUpperCase();

  fields.entryDate = firstDateNear(text, ["entry", "entered", "arrival", "i-94", "port of entry"]);
  fields.i130ApprovedDate = firstDateNear(text, ["i-130", "i130", "petition approved"]);
  fields.i485FiledDate = firstDateNear(text, ["i-485", "i485", "filed adjustment"]);

  fields.portOfEntry = matchGroup(
    /(?:port of entry|POE|entered (?:at|through))\s*[:\-]?\s*([A-Za-z\s,]+?)(?:\.|,|\n|$)/i,
    text,
  );
  fields.entryVisaType = matchGroup(
    /(?:visa type|entered on|admitted as)\s*[:\-]?\s*((?:B-?\d|F-?\d|H-?\d|[A-Z]-?\d)[^\n,.]{0,40})/i,
    text,
  );

  fields.applicantName = matchGroup(
    /(?:applicant|client|beneficiary)(?:'s)?\s*(?:name)?\s*[:\-]?\s*([A-Z][a-z]+(?:\s+[A-Z][a-z]+){1,3})/,
    text,
  );
  fields.petitionerName = matchGroup(
    /(?:petitioner|sponsor|usc\s+citizen)\s*[:\-]?\s*([A-Z][a-z]+(?:\s+[A-Z][a-z]+){0,3})/i,
    text,
  );

  const relMatch = text.match(
    /(?:relationship|petitioner is)\s*[:\-]?\s*(daughter|son|spouse|parent|mother|father|sibling)/i,
  );
  fields.petitionerRelationship = relMatch?.[1] ?? "";

  if (/hardship|extreme hardship/i.test(text)) {
    const m = text.match(/(?:extreme hardship|hardship)[^.]{0,400}\./i);
    fields.extremeHardshipFactors = (m?.[0] ?? "Hardship factors referenced in summary").trim();
  }
  if (/overstay|out of status/i.test(text)) {
    fields.clientStatus = "Out of status / overstay";
    const m = text.match(/(?:overstay|out of status)[^.]{0,300}\./i);
    fields.adverseFacts = (m?.[0] ?? "Overstay referenced in summary").trim();
    fields.adverseFactorBrief = "an overstay";
  }
  if (/autism|IEP|disabilit|special needs/i.test(text)) {
    const m = text.match(/(?:autism|IEP|disabilit|special needs)[^.]{0,400}\./i);
    if (m) fields.sectionAFacts = m[0].trim();
  }
  if (/positive|equit|family ties|community|tax compliance/i.test(text)) {
    const m = text.match(/(?:positive|equit|family ties|community)[^.]{0,300}\./i);
    fields.positiveEquities = (m?.[0] ?? "Positive equities referenced").trim();
  }
  if (/212\s*\(\s*a\s*\)|inadmissib/i.test(text)) {
    const m = text.match(/(?:INA\s*)?§?\s*212\s*\([^)]+\)[^.]{0,160}/i);
    fields.inadmissibilityGrounds = (m?.[0] ?? "INA §212(a) grounds referenced").trim();
  }
  if (/depart|consular processing|10.?year|3.?year bar/i.test(text)) {
    const m = text.match(/(?:depart|consular|bar|separat)[^.]{0,300}\./i);
    fields.departureHarm = (m?.[0] ?? "Departure harm referenced").trim();
  }

  fields.reliefSought = "Favorable exercise of discretion and approval of Form I-485";

  const filled = Object.values(fields).filter((v) => v.trim()).length;
  const confidence: AosExtractResult["confidence"] =
    filled >= 8 ? "medium" : filled >= 3 ? "low" : "low";

  return {
    fields,
    paragraphSelections: {},
    additionalNotes: "",
    confidence,
    extractionMode: "heuristic",
    llmAvailable: false,
  };
}

export function applyAosExtractToPayload(
  payload: DraftingFactsPayload,
  extract: AosExtractResult,
  options?: { overwrite?: boolean },
): { payload: DraftingFactsPayload; filledFieldIds: string[] } {
  const overwrite = options?.overwrite ?? false;
  const fields = { ...payload.fields };
  const filledFieldIds: string[] = [];

  for (const [key, value] of Object.entries(extract.fields)) {
    if (!value.trim()) continue;
    const current = fields[key];
    const hasValue = Array.isArray(current) ? current.length > 0 : Boolean(String(current ?? "").trim());
    if (hasValue && !overwrite) continue;
    fields[key] = value;
    filledFieldIds.push(key);
  }

  const paragraphSelections = { ...(payload.paragraphSelections ?? {}) };
  for (const [key, value] of Object.entries(extract.paragraphSelections)) {
    if (value) paragraphSelections[key] = value;
  }

  const additionalNotes = extract.additionalNotes.trim()
    ? [payload.additionalNotes, extract.additionalNotes].filter(Boolean).join("\n\n").trim()
    : payload.additionalNotes;

  return {
    payload: {
      ...payload,
      fields,
      paragraphSelections,
      additionalNotes,
      updatedAt: new Date().toISOString(),
    },
    filledFieldIds,
  };
}

/** Also fold generic OCR heuristics into AOS checklist fields. */
export function mergeSummaryIntoAosFacts(
  matterId: string,
  caseType: string,
  summary: string,
  existing?: DraftingFactsPayload | null,
): { payload: DraftingFactsPayload; filledFieldIds: string[] } {
  const base = existing ?? emptyDraftingFacts(matterId, caseType, "aos-discretionary-brief");
  const heuristic = heuristicExtractAosFacts(summary);
  const { payload: afterAos, filledFieldIds: aosFilled } = applyAosExtractToPayload(base, heuristic);

  const ocrFacts: OcrFact[] = [];
  const { payload, filledFieldIds: ocrFilled } = mergeOcrIntoDraftingFacts(afterAos, ocrFacts, summary);

  return {
    payload,
    filledFieldIds: [...new Set([...aosFilled, ...ocrFilled])],
  };
}

export async function extractAosFactsFromSummary(
  summary: string,
): Promise<AosExtractResult & { error?: string }> {
  try {
    const resp = await fetch("/api/aos/extract-facts", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ summary }),
    });
    const data = (await resp.json()) as AosExtractResult & { error?: string };
    if (!resp.ok) {
      const fallback = heuristicExtractAosFacts(summary);
      return { ...fallback, error: data.error ?? "Extract API failed — used heuristic mode." };
    }
    return data;
  } catch {
    return heuristicExtractAosFacts(summary);
  }
}

File: services/api/app/services/aos_fact_extract.py
"""AOS discretionary brief — extract structured drafting facts from attorney paste summary."""

from __future__ import annotations

import json
import re
from typing import Any

from app.services.llm import generate_text, is_configured

AOS_FIELD_IDS = (
    "applicantName",
    "aNumber",
    "clientStatus",
    "entryDate",
    "portOfEntry",
    "entryVisaType",
    "petitionerName",
    "petitionerRelationship",
    "qualifyingRelative",
    "i130ApprovedDate",
    "i485FiledDate",
    "reliefSought",
    "caseTheme",
    "caseThemeBrief",
    "sectionAHeading",
    "sectionAFacts",
    "sectionBHeading",
    "sectionBFacts",
    "adverseHeading",
    "adverseFactorBrief",
    "positiveEquities",
    "balancingInventory",
    "adverseFacts",
    "adverseFactors",
    "departureHarm",
    "extremeHardshipFactors",
    "inadmissibilityGrounds",
    "priorFilings",
)

EXTRACT_SYSTEM = """You are an immigration associate extracting facts for an AOS discretionary brief (I-485).
Given an attorney's client fact summary, return ONLY valid JSON mapping to structured fields.
Rules:
- Extract only what is stated or clearly implied — do not invent facts.
- Dates: ISO YYYY-MM-DD when possible, else preserve attorney's phrasing.
- caseTheme: one factual sentence with stakes (attorney will edit).
- sectionAHeading / sectionBHeading / adverseHeading: argument claims, not category labels.
- paragraphSelections: suggest library keys only when confident (section_a, section_d_adverse, section_e_balancing).
- Leave unknown fields as empty strings.
- additionalNotes: capture nuance not mapped to a field."""

EXTRACT_USER = """Attorney fact summary:
{summary}

Return JSON:
{{
  "fields": {{ "<field_id>": "<value or empty string>", ... }},
  "paragraphSelections": {{ "section_a": "<key.variant or empty>", ... }},
  "additionalNotes": "<optional string>",
  "confidence": "high|medium|low",
  "extraction_mode": "llm"
}}"""

DATE_PATTERNS = (
    re.compile(
        r"\b(?:Jan(?:uary)?|Feb(?:ruary)?|Mar(?:ch)?|Apr(?:il)?|May|Jun(?:e)?|Jul(?:y)?|Aug(?:ust)?|"
        r"Sep(?:tember)?|Oct(?:ober)?|Nov(?:ember)?|Dec(?:ember)?)\.?\s+\d{1,2},?\s+\d{4}\b",
        re.I,
    ),
    re.compile(r"\b\d{1,2}[/-]\d{1,2}[/-]\d{2,4}\b"),
    re.compile(r"\b\d{4}-\d{2}-\d{2}\b"),
)


def _parse_llm_json(raw: str) -> dict[str, Any] | None:
    text = raw.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text)
        text = re.sub(r"\s*```$", "", text)
    try:
        data = json.loads(text)
        return data if isinstance(data, dict) else None
    except json.JSONDecodeError:
        match = re.search(r"\{[\s\S]*\}", text)
        if not match:
            return None
        try:
            data = json.loads(match.group())
            return data if isinstance(data, dict) else None
        except json.JSONDecodeError:
            return None


def _first_date_near(text: str, keywords: tuple[str, ...]) -> str:
    lower = text.lower()
    for kw in keywords:
        idx = lower.find(kw)
        if idx < 0:
            continue
        window = text[max(0, idx - 80) : idx + len(kw) + 120]
        for pattern in DATE_PATTERNS:
            match = pattern.search(window)
            if match:
                return match.group().strip()
    return ""


def _match_group(pattern: str, text: str, flags: int = re.I) -> str:
    match = re.search(pattern, text, flags)
    return match.group(1).strip() if match else ""


def heuristic_extract_aos_facts(summary: str) -> dict[str, Any]:
    """Deterministic bucket extract when LLM unavailable."""
    text = summary.strip()
    fields: dict[str, str] = {fid: "" for fid in AOS_FIELD_IDS}
    if not text:
        return {
            "fields": fields,
            "paragraphSelections": {},
            "additionalNotes": "",
            "confidence": "low",
            "extraction_mode": "heuristic",
        }

    a_num = re.search(r"\bA-?\d{8,9}\b", text, re.I)
    if a_num:
        fields["aNumber"] = a_num.group().upper().replace("A", "A-") if "A-" not in a_num.group() else a_num.group()

    fields["entryDate"] = _first_date_near(text, ("entry", "entered", "arrival", "i-94", "port of entry"))
    fields["i130ApprovedDate"] = _first_date_near(text, ("i-130", "i130", "petition approved", "approved petition"))
    fields["i485FiledDate"] = _first_date_near(text, ("i-485", "i485", "filed adjustment", "adjustment application"))

    fields["portOfEntry"] = _match_group(
        r"(?:port of entry|POE|entered (?:at|through))\s*[:\-]?\s*([A-Za-z\s,]+?)(?:\.|,|\n|$)",
        text,
    )
    fields["entryVisaType"] = _match_group(
        r"(?:visa type|entered on|admitted as)\s*[:\-]?\s*((?:B-?\d|F-?\d|H-?\d|[A-Z]-?\d)[^\n,.]{0,40})",
        text,
    )

    fields["applicantName"] = _match_group(
        r"(?:applicant|client|beneficiary|respondent)(?:'s)?\s*(?:name)?\s*[:\-]?\s*([A-Z][a-z]+(?:\s+[A-Z][a-z]+){1,3})",
        text,
    )
    fields["petitionerName"] = _match_group(
        r"(?:petitioner|sponsor|usc\s+citizen)\s*[:\-]?\s*([A-Z][a-z]+(?:\s+[A-Z][a-z]+){0,3})",
        text,
    )
    fields["petitionerRelationship"] = _match_group(
        r"(?:relationship|petitioner is (?:her|his|their)?)\s*[:\-]?\s*(daughter|son|spouse|parent|mother|father|sibling|brother|sister)",
        text,
    )
    if not fields["petitionerRelationship"]:
        for rel in ("daughter", "son", "spouse", "parent", "mother", "father"):
            if re.search(rf"\b{rel}\b", text, re.I):
                fields["petitionerRelationship"] = rel
                break

    fields["qualifyingRelative"] = _match_group(
        r"(?:qualifying relative|hardship to)\s*[:\-]?\s*([^\n,.]{5,80})",
        text,
    )

    if re.search(r"\bout of status\b|\boverstay\b|\bvisa overstay\b|\bunlawful presence\b", text, re.I):
        fields["clientStatus"] = "Out of status / overstay"
    elif re.search(r"\btps\b", text, re.I):
        fields["clientStatus"] = "TPS"
    elif re.search(r"\blpr\b|\bpermanent resident\b", text, re.I):
        fields["clientStatus"] = "Lawful permanent resident"
    elif re.search(r"\bpending i-485\b|\bpending adjustment\b", text, re.I):
        fields["clientStatus"] = "Pending I-485"

    if re.search(r"\bhardship\b|\bextreme hardship\b", text, re.I):
        m = re.search(r"(?:extreme hardship|hardship)[^.]{0,400}\.", text, re.I)
        fields["extremeHardshipFactors"] = (m.group(0) if m else "Hardship factors referenced in summary").strip()

    if re.search(r"212\s*\(\s*a\s*\)|inadmissib", text, re.I):
        m = re.search(r"(?:INA\s*)?§?\s*212\s*\([^)]+\)[^.]{0,160}", text, re.I)
        fields["inadmissibilityGrounds"] = (m.group(0) if m else "INA §212(a) grounds referenced").strip()

    if re.search(r"\bpositive\b|\bequit|\bfamily ties\b|\bcommunity\b|\btax compliance\b", text, re.I):
        m = re.search(r"(?:positive|equit|family ties|community)[^.]{0,300}\.", text, re.I)
        fields["positiveEquities"] = (m.group(0) if m else "Positive equities referenced in summary").strip()

    if re.search(r"\boverstay\b|\bcriminal\b|\bdenial\b|\bviolation\b|\badverse\b", text, re.I):
        m = re.search(r"(?:overstay|adverse|criminal|denial|violation)[^.]{0,300}\.", text, re.I)
        fields["adverseFacts"] = (m.group(0) if m else "Adverse factors referenced in summary").strip()
        fields["adverseFactorBrief"] = "an overstay" if re.search(r"\boverstay\b", text, re.I) else ""

    if re.search(r"\bdepart|\bconsular processing\b|\b3.?year\b|\b10.?year\b|\bbar\b", text, re.I):
        m = re.search(r"(?:depart|consular|bar|separat)[^.]{0,300}\.", text, re.I)
        fields["departureHarm"] = (m.group(0) if m else "Departure/consular harm referenced").strip()

    if re.search(r"\bautism\b|\biep\b|\bdisabilit|\bspecial needs\b", text, re.I):
        m = re.search(r"(?:autism|IEP|disabilit|special needs)[^.]{0,400}\.", text, re.I)
        fields["sectionAFacts"] = (m.group(0) if m else "").strip()

    if re.search(r"\bprior (?:filing|application|petition)\b|\bprevious (?:a|i)-\d", text, re.I):
        m = re.search(r"(?:prior|previous)[^.]{0,300}\.", text, re.I)
        fields["priorFilings"] = (m.group(0) if m else "Prior immigration history referenced").strip()

    fields["reliefSought"] = "Favorable exercise of discretion and approval of Form I-485"

    filled = sum(1 for v in fields.values() if v.strip())
    confidence = "medium" if filled >= 6 else "low" if filled >= 2 else "low"

    return {
        "fields": fields,
        "paragraphSelections": {},
        "additionalNotes": "",
        "confidence": confidence,
        "extraction_mode": "heuristic",
    }


def extract_aos_facts(summary: str, *, force_heuristic: bool = False) -> dict[str, Any]:
    """LLM extract when Anthropic configured; else heuristic."""
    trimmed = summary.strip()
    if not trimmed:
        return heuristic_extract_aos_facts("")

    if force_heuristic or not is_configured():
        return heuristic_extract_aos_facts(trimmed)

    raw = generate_text(
        system=EXTRACT_SYSTEM,
        user=EXTRACT_USER.format(summary=trimmed[:12000]),
        max_tokens=4096,
        temperature=0.1,
    )
    if not raw:
        result = heuristic_extract_aos_facts(trimmed)
        result["llm_fallback"] = True
        return result

    parsed = _parse_llm_json(raw)
    if not parsed:
        result = heuristic_extract_aos_facts(trimmed)
        result["llm_fallback"] = True
        return result

    fields_in = parsed.get("fields") if isinstance(parsed.get("fields"), dict) else {}
    fields: dict[str, str] = {fid: "" for fid in AOS_FIELD_IDS}
    for key, val in fields_in.items():
        if key in fields and val is not None:
            fields[key] = str(val).strip()

    selections_raw = parsed.get("paragraphSelections")
    selections: dict[str, str] = {}
    if isinstance(selections_raw, dict):
        for k, v in selections_raw.items():
            if v and isinstance(v, str):
                selections[str(k)] = v.strip()

    return {
        "fields": fields,
        "paragraphSelections": selections,
        "additionalNotes": str(parsed.get("additionalNotes") or "").strip(),
        "confidence": str(parsed.get("confidence") or "medium"),
        "extraction_mode": "llm",
    }
